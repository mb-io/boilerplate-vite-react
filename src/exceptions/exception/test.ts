import { describe, it, expect } from 'vitest'
import { Exception } from './exception'

describe('Exception', () => {
  it('should create an exception with a default name and message', () => {
    const exception = new Exception()
    expect(exception.name).toBe('UnknownException')
    expect(exception.message).toBe('')
    expect(exception).toMatchSnapshot()
  })

  it('should create an exception with a custom message', () => {
    const message = 'Custom error message'
    const exception = new Exception(message)
    expect(exception.name).toBe('UnknownException')
    expect(exception.message).toBe(message)
    expect(exception).toMatchSnapshot()
  })
})
