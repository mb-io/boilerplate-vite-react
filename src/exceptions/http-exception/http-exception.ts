import { AxiosError } from 'axios'

export class HttpException extends Error {
  status: number
  message: string
  error?: AxiosError

  constructor(status: number, message: string, error?: AxiosError) {
    super(message)
    this.status = status
    this.message = message
    this.error = error
  }
}
