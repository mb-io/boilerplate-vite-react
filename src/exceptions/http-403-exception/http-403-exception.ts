import { AxiosError } from 'axios'

import { HttpException } from '../http-exception/http-exception'

export class Http403Exception extends HttpException {
  constructor(message?: string, error?: AxiosError) {
    super(403, message ?? 'Not Authorized', error)
  }
}
