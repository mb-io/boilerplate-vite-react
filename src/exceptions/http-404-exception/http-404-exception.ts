import { AxiosError } from 'axios'

import { HttpException } from '../http-exception/http-exception'

export class Http404Exception extends HttpException {
  constructor(message?: string, error?: AxiosError) {
    super(404, message ?? 'Not Found', error)
  }
}
