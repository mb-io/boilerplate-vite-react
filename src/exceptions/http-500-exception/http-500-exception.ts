import { AxiosError } from 'axios'

import { HttpException } from '../http-exception/http-exception'

export class Http500Exception extends HttpException {
  constructor(message?: string, error?: AxiosError) {
    super(500, message ?? 'Internal Server Error', error)
  }
}
