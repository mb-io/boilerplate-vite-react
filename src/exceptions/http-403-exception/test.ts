import { describe, it, expect } from 'vitest'
import { Http403Exception } from './http-403-exception'
import { AxiosError } from 'axios'

describe('Http403Exception', () => {
  it('should create an exception with default message', () => {
    const exception = new Http403Exception()
    expect(exception.status).toBe(403)
    expect(exception.message).toBe('Not Authorized')
    expect(exception).toMatchSnapshot()
  })

  it('should create an exception with custom message', () => {
    const message = 'Custom not authorized message'
    const exception = new Http403Exception(message)
    expect(exception.status).toBe(403)
    expect(exception.message).toBe(message)
    expect(exception).toMatchSnapshot()
  })

  it('should create an exception with custom message and AxiosError', () => {
    const message = 'Custom not authorized message'
    const axiosError = new AxiosError('Axios error message')
    const exception = new Http403Exception(message, axiosError)
    expect(exception.status).toBe(403)
    expect(exception.message).toBe(message)
    expect(exception.error).toBe(axiosError)
    expect(exception).toMatchSnapshot()
  })
})
