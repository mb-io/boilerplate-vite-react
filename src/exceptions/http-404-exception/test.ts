import { describe, it, expect } from 'vitest'
import { Http404Exception } from './http-404-exception'
import { AxiosError } from 'axios'

describe('Http404Exception', () => {
  it('should create an exception with default message', () => {
    const exception = new Http404Exception()
    expect(exception.status).toBe(404)
    expect(exception.message).toBe('Not Found')
    expect(exception).toMatchSnapshot()
  })

  it('should create an exception with custom message', () => {
    const message = 'Custom not found message'
    const exception = new Http404Exception(message)
    expect(exception.status).toBe(404)
    expect(exception.message).toBe(message)
    expect(exception).toMatchSnapshot()
  })

  it('should create an exception with custom message and AxiosError', () => {
    const message = 'Custom not found message'
    const axiosError = new AxiosError('Axios error message')
    const exception = new Http404Exception(message, axiosError)
    expect(exception.status).toBe(404)
    expect(exception.message).toBe(message)
    expect(exception.error).toBe(axiosError)
    expect(exception).toMatchSnapshot()
  })
})
