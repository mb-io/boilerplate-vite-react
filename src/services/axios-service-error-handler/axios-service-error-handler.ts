import { AxiosError } from 'axios'

import { Exception } from '@/exceptions/exception/exception'
import { Http403Exception } from '@/exceptions/http-403-exception/http-403-exception'
import { Http404Exception } from '@/exceptions/http-404-exception/http-404-exception'
import { Http500Exception } from '@/exceptions/http-500-exception/http-500-exception'
import { HttpException } from '@/exceptions/http-exception/http-exception'
import { HttpNoResponseException } from '@/exceptions/http-no-response-exception/http-no-response-exception'

export const axiosServiceErrorHandler = (error: AxiosError, message?: string) => {
  if (error.response) {
    // The request was made and the server responded with a status code
    // that falls out of the range of 2xx
    const status = error.response.status

    switch (true) {
      case status === 404:
        throw new Http404Exception(message, error)
      case status === 403:
        throw new Http403Exception(message, error)
      case status >= 500 && status <= 599:
        throw new Http500Exception(error.message, error)
      default:
        throw new HttpException(error.response.status, error.message, error)
    }
  } else if (error.request) {
    // The request was made but no response was received
    // `error.request` is an instance of XMLHttpRequest in the browser and an instance of
    // http.ClientRequest in node.js
    throw new HttpNoResponseException(error.message)
  } else {
    // Something happened in setting up the request that triggered an Error
    throw new Exception(error.message)
  }
}
