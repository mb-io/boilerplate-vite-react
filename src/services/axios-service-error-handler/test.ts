import { describe, it, expect } from 'vitest'
import { AxiosError, AxiosHeaders } from 'axios'
import { axiosServiceErrorHandler } from './axios-service-error-handler'
import { Http403Exception } from '@/exceptions/http-403-exception/http-403-exception'
import { Http404Exception } from '@/exceptions/http-404-exception/http-404-exception'
import { Http500Exception } from '@/exceptions/http-500-exception/http-500-exception'
import { HttpException } from '@/exceptions/http-exception/http-exception'
import { Exception } from '@/exceptions/exception/exception'

describe('axiosServiceErrorHandler', () => {
  it('should throw Http404Exception for 404 status', () => {
    const error = new AxiosError('Not Found', undefined, undefined, undefined, {
      status: 404,
      data: {},
      headers: {},
      statusText: 'Not Found',
      config: {
        headers: {} as AxiosHeaders
      }
    })
    expect(() => axiosServiceErrorHandler(error)).toThrow(Http404Exception)
    expect(() => axiosServiceErrorHandler(error)).toMatchSnapshot()
  })

  it('should throw Http403Exception for 403 status', () => {
    const error = new AxiosError('Not Authorized', undefined, undefined, undefined, {
      status: 403,
      data: {},
      headers: {},
      statusText: 'Not Authorized',
      config: {
        headers: {} as AxiosHeaders
      }
    })
    expect(() => axiosServiceErrorHandler(error)).toThrow(Http403Exception)
    expect(() => axiosServiceErrorHandler(error)).toMatchSnapshot()
  })

  it('should throw Http500Exception for 500 status', () => {
    const error = new AxiosError('Internal Server Error', undefined, undefined, undefined, {
      status: 500,
      data: {},
      headers: {},
      statusText: 'Internal Server Error',
      config: {
        headers: {} as AxiosHeaders
      }
    })
    expect(() => axiosServiceErrorHandler(error)).toThrow(Http500Exception)
    expect(() => axiosServiceErrorHandler(error)).toMatchSnapshot()
  })

  it('should throw HttpException for other status', () => {
    const error = new AxiosError('Bad Request', undefined, undefined, undefined, {
      status: 400,
      data: {},
      headers: {},
      statusText: 'Bad Request',
      config: {
        headers: {} as AxiosHeaders
      }
    })
    expect(() => axiosServiceErrorHandler(error)).toThrow(HttpException)
    expect(() => axiosServiceErrorHandler(error)).toMatchSnapshot()
  })

  it('should throw Exception for other errors', () => {
    const error = new AxiosError('Unknown Error')
    expect(() => axiosServiceErrorHandler(error)).toThrow(Exception)
    expect(() => axiosServiceErrorHandler(error)).toMatchSnapshot()
  })
})
