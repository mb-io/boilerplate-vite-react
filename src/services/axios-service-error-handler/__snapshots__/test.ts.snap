// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`axiosServiceErrorHandler > should throw Exception for other errors 1`] = `[Function]`;

exports[`axiosServiceErrorHandler > should throw Http403Exception for 403 status 1`] = `[Function]`;

exports[`axiosServiceErrorHandler > should throw Http404Exception for 404 status 1`] = `[Function]`;

exports[`axiosServiceErrorHandler > should throw Http500Exception for 500 status 1`] = `[Function]`;

exports[`axiosServiceErrorHandler > should throw HttpException for other status 1`] = `[Function]`;
